featurebyte/query_graph/transform/offline_ingest_extractor.py

```diff
@@ -0,0 +1,319 @@
+        """
+        Update primary entity IDs for the given node name
+
+        Parameters
+        ----------
+        node: Node
+            Node to be processed
+        input_node_names: List[str]
+            List of input node names
+        """
+        primary_entity_ids: List[ObjectId]
+        if isinstance(node.parameters, BaseGroupbyParameters):
+            # primary entity ids introduced by groupby node family
+            primary_entity_ids = node.parameters.entity_ids or []  # type: ignore
+        elif isinstance(node, (LookupNode, LookupTargetNode)):
+            # primary entity ids introduced by lookup node family
+            primary_entity_ids = [node.parameters.entity_id]
+        else:
+            # primary entity ids inherited from input nodes
+            primary_entity_ids = []
+            for input_node_name in input_node_names:
+                primary_entity_ids.extend(self.node_name_to_primary_entity_ids[input_node_name])
+            primary_entity_ids = list(set(primary_entity_ids))
+
+        # reduce the primary entity ids based on entity relationship
+        primary_entity_ids = self.entity_ancestor_descendant_mapper.reduce_entity_ids(
+            entity_ids=list(primary_entity_ids)
+        )
+        self.node_name_to_primary_entity_ids[node.name] = sorted(primary_entity_ids)
+
+    def should_decompose_query_graph(self, node_name: str, input_node_names: List[str]) -> bool:
+        """
+        Check whether to decompose the query graph into graph with nested offline store ingest query nodes
+
+        Parameters
+        ----------
+        node_name: str
+            Node name
+        input_node_names: List[str]
+            List of input node names
+
+        Returns
+        -------
+        bool
+        """
+        if not input_node_names:
+            # if there is no input node, that means this is the starting node
+            return False
+
+        output_entity_ids = self.node_name_to_primary_entity_ids[node_name]
+        all_input_entity_are_empty = True
+        for input_node_name in input_node_names:
+            input_entity_ids = self.node_name_to_primary_entity_ids[input_node_name]
+            if input_entity_ids == output_entity_ids:
+                # if any of the input is the same as the output, that means no new entity ids are added
+                # to the universe. So we should not split the query graph.
+                return False
+            if input_entity_ids:
+                all_input_entity_are_empty = False
+
+        if all_input_entity_are_empty:
+            # if all the input nodes are empty, that means the output node is the starting node
+            # that introduces new entity ids to the universe. So we should not split the query graph.
+            return False
+
+        # if none of the above conditions are met, that means we should split the query graph
+        return True
```

Comment:
The decomposition should also be done if on demand columns such as point in time are used by the feature. Is that handled by the current logic?

Reply:
Current logic does not consider request column. Will update the extractor to handle request column.

Reply:
This method is updated to consider request column node. Tests are added to check query graph with request column node(s).